api/envoy/config/core/v3/address.proto

```diff
@@ -30,6 +30,16 @@ message Pipe {
   uint32 mode = 2 [(validate.rules).uint32 = {lte: 511}];
 }
 
+// The address represents an envoy internal listener.
+message EnvoyInternalAddress {
```

Comment:
Thanks for starting to break things out!

Sorry, I thought I called this out on the larger PR (failing to find my comment there) but why do we need a new address type?  Functionally if you multiplex a bunch of requests over a single connection the underlying connection is still one of the regular address types (comes in on a socket or pipe) and I think we could just expose that as the address of the connection.
Before I dive into code let's sort out the pros and cons of the new address type vs just exposing the actual address of the connection

Reply:
> Sorry, I thought I called this out on the larger PR (failing to find my comment there) but why do we need a new address type? Functionally if you multiplex a bunch of requests over a single connection the underlying connection is still one of the regular address types (comes in on a socket or pipe)

Sorry I didn't looked into the large PR... That PR is supposed to provide only the big picture.

The underlying connection is still using the regular address types.
The internal address is used by the HCM de-plex the request. 
In my case, the upstream cluster of the de-plex request is always another listener beside the HCM in the same envoy.

An OS pipe, an socket pair, or even a tcp to 127.0.0.1 works in term of connectivity. However, performance suffers.

The large PR is addressing the implementation of an envoy-intenral-connection in which bytes doesn't go through kernel.

But another question is, how and when do we choose to use the envoy-internal-connection?

This PR is my proposed solution: 
we can introduce a new address type.

The end 2 end flow is
1. HCM de-plex the h2 request, matching a cluster C
2. The cluster C has the endpoint with the internal address in this PR
3. The the cluster establish the connection to the new address, create internal-connection(Dispatcher::createPipeConnection) instead of traditional tcp connection(Dispatcher::createClientConnection).


Alternatively the magic could be hidden by createClientConnection, something like 
if the address is 127.0.0.1:9000 and envoy happens to listen on 127.0.0.1:9000, then create internal connection instead of tcp connection.

I personally think it's much clearer and deterministic if we introduce new internal address type as well as new internal listener type in https://github.com/envoyproxy/envoy/pull/12838/files#r480355785

Reply:
Yeah, I guess the way I envision it, it a given network connection comes in on listener Foo with address bar, Envoy may have to create a "virtual" listener, but it shouldn't have to be hand-configured.  The filter chains on Foo should apply to the demuxed connection, and any L4 or L7 sessions should report having come in on address bar.

I can absolutely see the utility of allowing configuring a separate listener chain with separate filters for the demuxed connection, though I think I'd prefer to see that work done later once we have a feel for how the whole system hangs together.  I'm quite unclear as to the use of the mocked out address - hopefully will get clarity when we chat later today.

Reply:
ah, sorted offline.  The address is *just* used for mapping, not reported by the connection.  SGTM.